File: 496-Next-Greater-Element-I.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer,Integer> idx = new HashMap<>();
        for(int i=0;i<nums1.length;i++){
            idx.put(nums1[i],i);
        }

        int[] ans = new int[nums1.length];
        Arrays.fill(ans,-1);
        
        Stack<Integer> res = new Stack<>();

        for(int i =0;i<nums2.length;i++){
            int curr = nums2[i];
            while(!res.isEmpty() && curr > res.peek()){
                int val = res.pop();
                int index = idx.get(val);
                ans[index] = curr;
            }
            if(idx.containsKey(curr)){
                res.push(curr);
            }
        }
        return ans;
    }
}

File: 496-Next-Greater-Element-I.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        idx = {n:i for i, n in enumerate(nums1)}
        ans = [-1] * len(nums1)
        stack = []

        for i in range(len(nums2)):
            curr = nums2[i]
            while stack and curr > stack[-1]:
                val = stack.pop()
                index = idx[val]
                ans[index] = curr
            if curr in nums1:
                stack.append(curr)
        return ans